rust/rope/src/tree.rs

```diff
@@ -478,7 +471,29 @@ impl<N: NodeInfo> TreeBuilder<N> {
     pub fn push(&mut self, n: Node<N>) {
         match self.0.take() {
             None => self.0 = Some(n),
-            Some(buf) => self.0 = Some(Node::concat(buf, n))
+            Some(buf) => self.0 = Some(Node::concat(buf, n)),
+        }
+    }
+
+    pub fn push_leaves(&mut self, leaves: Vec<N::L>) {
+        let mut stack: Vec<Vec<Node<N>>> = Vec::new();
+        for leaf in leaves {
```

Comment:
I found this a bit hard to follow (and I probably wrote the original logic). What I think would be most helpful is a description of invariants. I _believe_ the following is true, and could be adapted into comments:

The stack is a list of list of nodes. Each list of nodes is of uniform height, and of length at least one and at most MAX_CHILDREN - 1. The heights are strictly decreasing.

Having gone through it, this looks right.

Reply:
Those invariants are correct. I added a description of those.

I was/am doubting between writing just a short doc string documenting the invariants and a full blown explanation why these invariants lead to amortized linear time.

Reply:
More detailed documentation almost never hurts. If you're not sure I'd say write both, much like you would write a long git commit message, i.e. a short oneliner, an empty line and a long description.

Reply:
Aaaand I see that's exactly what you did. These github change views can be confusing at times.